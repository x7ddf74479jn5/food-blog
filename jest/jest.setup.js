// Optional: configure or set up a testing framework before each test.
// If you delete this file, remove `setupFilesAfterEnv` from `jest.config.js`

// Used for __tests__/testing-library.js
// Learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom/extend-expect";

window.matchMedia = jest.fn().mockImplementation((query) => {
  return {
    addListener: jest.fn(),
    matches: false,
    media: query,
    onchange: null,
    removeListener: jest.fn(),
  };
});
