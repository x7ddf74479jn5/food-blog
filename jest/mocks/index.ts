export { mockNextImage } from "./image";
export type { NextRouter } from "./router";
export { mockRouter, RouterContext, withMockedRouter } from "./router";
export { setupWindow } from "./window";
