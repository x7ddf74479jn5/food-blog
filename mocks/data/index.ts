export { mockArticles } from "./articles";
export { mockCategories } from "./categories";
export { mockConfig } from "./config";
export { mockPickup } from "./pickup";
export { mockTags } from "./tags";
export { dateCommon } from "./utils";
export { mockWriters } from "./writers";
