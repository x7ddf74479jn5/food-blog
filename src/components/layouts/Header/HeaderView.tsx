import { SiteTitle } from "@/components/atoms/SiteTitle";
import { CategoryMenu } from "@/components/model/category/CategoryMenu";
import { ThemeSwitch } from "@/components/molecules/ThemeSwitch";
import { SearchArea } from "@/components/feature/SearchArea";
import type { TCategory, TTag } from "@/types";

type Props = {
  categories: TCategory[];
  tags: TTag[];
};

export const HeaderView: React.FC<Props> = ({ categories, tags }) => {
  return (
    <header className="py-2">
      <div className="grid auto-rows-auto grid-cols-5">
        <div className="col-start-1 col-end-3 justify-self-start lg:col-end-2">
          {/* @ts-expect-error Server Component */}
          <SiteTitle />
        </div>
        <div className="col-span-full row-start-2 mt-4 md:col-span-3 md:col-start-2 md:col-end-5 lg:row-start-1 lg:mt-0">
          <SearchArea categories={categories} tags={tags} />
        </div>
        <div className="col-start-3 col-end-6 flex items-center space-x-4 self-start justify-self-end lg:col-start-5">
          <ThemeSwitch />
          <CategoryMenu categories={categories} />
        </div>
      </div>
    </header>
  );
};
