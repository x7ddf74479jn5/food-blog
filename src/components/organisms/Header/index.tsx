import { memo } from "react";

import { SiteTitle } from "@/components/atoms/SiteTitle";
import { CategoryMenu } from "@/components/molecules/category/CategoryMenu";
import { ThemeSwitch } from "@/components/molecules/ThemeSwitch";
import { SearchArea } from "@/components/organisms/SearchArea";
import type { TCategory, TTag } from "@/types";

type HeaderProps = {
  siteTitle: string;
  categories: TCategory[];
  tags: TTag[];
};

const Header: React.FC<HeaderProps> = ({ siteTitle, categories, tags }) => {
  return (
    <header className="py-2">
      <div className="flex items-start justify-between">
        <SiteTitle size="text-2xl" title={siteTitle} />
        <div className="hidden w-1/2 sm:block" aria-label="search-area-pc">
          <SearchArea categories={categories} tags={tags} />
        </div>
        <div className="flex items-center space-x-4">
          <ThemeSwitch />
          <CategoryMenu categories={categories} />
        </div>
      </div>
      <div className="mt-2 block sm:hidden" aria-label="search-area-mobile">
        <SearchArea categories={categories} tags={tags} />
      </div>
    </header>
  );
};

export default memo(Header);
