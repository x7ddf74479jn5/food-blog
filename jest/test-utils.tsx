import type Queries from "@testing-library/dom/types/queries";
import type { RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";
import type { RequestHandler } from "msw";
import { setupServer } from "msw/node";
import { SWRConfig } from "swr";

import { SearchProvider } from "@/contexts/search/SearchContext";
import { ViewportProvider } from "@/contexts/viewport/ViewportContext";

import { withMockLegacyRouter, withMockRouter } from "./mocks";

export const Providers: React.ComponentType<{ children?: React.ReactNode }> = ({ children }) => {
  return withMockRouter(
    withMockLegacyRouter(
      <SWRConfig value={{ dedupingInterval: 0, provider: () => new Map() }}>
        <SearchProvider>
          <ViewportProvider>{children}</ViewportProvider>
        </SearchProvider>
      </SWRConfig>
    )
  );
};

const customRender = (ui: React.ReactElement, options = {}): RenderResult<typeof Queries, HTMLElement> => {
  return render(ui, { wrapper: Providers, ...options });
};

// re-export everything
export * from "./mocks";
export * from "@testing-library/react";

// override render method
export { render as originalRender, customRender as render };

export const reTestCase = {
  anyImage: expect.stringMatching(/^(data:image\/gif)|\.(png|webp|jpeg|jpg|svg)$/),
  anyWord: expect.stringMatching(/\w+/),
};

type ErrorComponentProps = {
  message?: string;
};
export const ErrorComponent: React.FC<ErrorComponentProps> = ({ message }) => {
  throw new Error(message);
};

export const setupMockServer = (...handlers: RequestHandler[]) => {
  const server = setupServer(...handlers);
  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());
  return server;
};
