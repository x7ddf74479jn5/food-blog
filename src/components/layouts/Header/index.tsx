import { fetchTags } from "@/repositories";
import { getCategories } from "@/services/category";

import { HeaderView } from "./HeaderView";

export const Header = async () => {
  const [categories, tags] = await Promise.all([getCategories(), fetchTags()]);

  return <HeaderView categories={categories} tags={tags} />;
};
