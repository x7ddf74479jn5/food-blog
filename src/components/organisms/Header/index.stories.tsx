import type { ComponentMeta, ComponentStoryObj } from "@storybook/react";
import { mockCategories } from "mocks/data";

import Header from ".";
import { withContext } from ".storybook/mocks/context";
import { screenshotViewportVariants } from ".storybook/screenshot";

export default {
  component: Header,
  title: "organisms/Header",
} as ComponentMeta<typeof Header>;

export const Default: ComponentStoryObj<typeof Header> = {
  args: {
    categories: Object.values(mockCategories),
    siteTitle: "Title",
  },
  decorators: [
    (storyFn) => {
      return withContext(storyFn());
    },
  ],
  parameters: {
    screenshot: {
      variants: screenshotViewportVariants,
    },
  },
};
