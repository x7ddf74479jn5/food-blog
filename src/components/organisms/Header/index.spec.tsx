import { mockCategories, mockConfig } from "@mocks/data";
import { render, screen } from "jest/test-utils";
import renderer from "react-test-renderer";

import Header from "../Header";

describe("components/organisms/Header", () => {
  const { siteTitle } = mockConfig;
  const categoryList = Object.values(mockCategories);

  it("snapshot", () => {
    const tree = renderer.create(<Header categories={categoryList} siteTitle={siteTitle} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("OK: 初期表示が正しい", () => {
    render(<Header categories={categoryList} siteTitle={siteTitle} />);
    const header = screen.getByRole("banner");
    expect(header).toBeInTheDocument();
    expect(header.tagName).toBe("HEADER");
    const anchor = screen.getByRole("link");
    expect(anchor).toHaveTextContent(siteTitle);
    expect(anchor).toHaveAttribute("href", "/");
    const searchBoxes = screen.getAllByRole("searchbox");
    expect(searchBoxes.length).toBe(2);
  });
});
