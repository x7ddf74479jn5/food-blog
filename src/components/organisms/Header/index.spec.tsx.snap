// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/Header snapshot 1`] = `
<header
  className="py-2"
>
  <div
    className="flex justify-between items-center"
  >
    <a
      className="text-2xl font-bold text-green-600"
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Food Blog
    </a>
    <div
      className="hidden sm:block w-1/3"
    >
      <div
        className="relative"
      >
        <div
          className="flex items-center py-2 px-3 space-x-2 bg-gray-100 dark:bg-gray-700 rounded-full"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "gray",
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
          <input
            className="w-full dark:text-gray-100 bg-gray-100 dark:bg-gray-700 focus:outline-none"
            inputMode="text"
            onChange={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            onSubmit={[Function]}
            placeholder="Search..."
            type="search"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="flex items-center space-x-4"
    >
      <div
        style={
          Object {
            "MozTransition": "opacity 0.25s",
            "MozUserSelect": "none",
            "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
            "WebkitTransition": "opacity 0.25s",
            "WebkitUserSelect": "none",
            "borderRadius": 12,
            "direction": "ltr",
            "display": "inline-block",
            "msUserSelect": "none",
            "opacity": 1,
            "position": "relative",
            "textAlign": "left",
            "touchAction": "none",
            "transition": "opacity 0.25s",
            "userSelect": "none",
          }
        }
      >
        <div
          className="react-switch-bg"
          onClick={[Function]}
          onMouseDown={[Function]}
          style={
            Object {
              "MozTransition": "background 0.25s",
              "WebkitTransition": "background 0.25s",
              "background": "#555555",
              "borderRadius": 12,
              "cursor": "pointer",
              "height": 24,
              "margin": 0,
              "position": "relative",
              "transition": "background 0.25s",
              "width": 48,
            }
          }
        >
          <div
            style={
              Object {
                "MozTransition": "opacity 0.25s",
                "WebkitTransition": "opacity 0.25s",
                "height": 24,
                "opacity": 0,
                "pointerEvents": "none",
                "position": "relative",
                "transition": "opacity 0.25s",
                "width": 27,
              }
            }
          >
            <div
              className="flex justify-center items-center h-full"
            >
              <svg
                fill="currentColor"
                height="80%"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": "yellow",
                  }
                }
                viewBox="0 0 512 512"
                width="80%"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M283.211 512c78.962 0 151.079-35.925 198.857-94.792 7.068-8.708-.639-21.43-11.562-19.35-124.203 23.654-238.262-71.576-238.262-196.954 0-72.222 38.662-138.635 101.498-174.394 9.686-5.512 7.25-20.197-3.756-22.23A258.156 258.156 0 0 0 283.211 0c-141.309 0-256 114.511-256 256 0 141.309 114.511 256 256 256z"
                />
              </svg>
            </div>
          </div>
          <div
            style={
              Object {
                "MozTransition": "opacity 0.25s",
                "WebkitTransition": "opacity 0.25s",
                "height": 24,
                "opacity": 1,
                "pointerEvents": "none",
                "position": "absolute",
                "right": 0,
                "top": 0,
                "transition": "opacity 0.25s",
                "width": 27,
              }
            }
          >
            <div
              className="flex justify-center items-center h-full"
            >
              <svg
                fill="currentColor"
                height="80%"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": "gold",
                  }
                }
                viewBox="0 0 512 512"
                width="80%"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 160c-52.9 0-96 43.1-96 96s43.1 96 96 96 96-43.1 96-96-43.1-96-96-96zm246.4 80.5l-94.7-47.3 33.5-100.4c4.5-13.6-8.4-26.5-21.9-21.9l-100.4 33.5-47.4-94.8c-6.4-12.8-24.6-12.8-31 0l-47.3 94.7L92.7 70.8c-13.6-4.5-26.5 8.4-21.9 21.9l33.5 100.4-94.7 47.4c-12.8 6.4-12.8 24.6 0 31l94.7 47.3-33.5 100.5c-4.5 13.6 8.4 26.5 21.9 21.9l100.4-33.5 47.3 94.7c6.4 12.8 24.6 12.8 31 0l47.3-94.7 100.4 33.5c13.6 4.5 26.5-8.4 21.9-21.9l-33.5-100.4 94.7-47.3c13-6.5 13-24.7.2-31.1zm-155.9 106c-49.9 49.9-131.1 49.9-181 0-49.9-49.9-49.9-131.1 0-181 49.9-49.9 131.1-49.9 181 0 49.9 49.9 49.9 131.1 0 181z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          className="react-switch-handle"
          onClick={[Function]}
          onMouseDown={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "MozTransition": "background-color 0.25s, transform 0.25s, box-shadow 0.15s",
              "WebkitTransition": "background-color 0.25s, transform 0.25s, box-shadow 0.15s",
              "background": "#ffffff",
              "border": 0,
              "borderRadius": "50%",
              "boxShadow": null,
              "cursor": "pointer",
              "display": "inline-block",
              "height": 20,
              "outline": 0,
              "position": "absolute",
              "top": 2,
              "transform": "translateX(2px)",
              "transition": "background-color 0.25s, transform 0.25s, box-shadow 0.15s",
              "width": 20,
            }
          }
        />
        <input
          aria-checked={false}
          aria-label="switch between day and night themes"
          checked={false}
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyUp={[Function]}
          role="switch"
          style={
            Object {
              "border": 0,
              "clip": "rect(0 0 0 0)",
              "height": 1,
              "margin": -1,
              "overflow": "hidden",
              "padding": 0,
              "position": "absolute",
              "width": 1,
            }
          }
          type="checkbox"
        />
      </div>
      <div
        className="inline-block relative text-left"
      >
        <span
          className="rounded-md shadow-sm"
        >
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="inline-flex justify-center py-1 px-3 w-full text-sm font-medium leading-5 text-gray-700 hover:text-gray-500 dark:text-gray-100 bg-white dark:bg-gray-700 rounded-md border border-gray-300 dark:border-gray-500 dark:hover:border-gray-400 transition duration-150 ease-in-out focus:outline-none"
            id="headlessui-menu-button-1"
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            type="button"
          >
            <span>
              カテゴリー
            </span>
            <svg
              className="-mr-1 ml-2 w-5 h-5"
              fill="currentColor"
              viewBox="0 0 20 20"
            >
              <path
                clipRule="evenodd"
                d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
                fillRule="evenodd"
              />
            </svg>
          </button>
        </span>
      </div>
    </div>
  </div>
  <div
    className="block sm:hidden mt-2"
  >
    <div
      className="relative"
    >
      <div
        className="flex items-center py-2 px-3 space-x-2 bg-gray-100 dark:bg-gray-700 rounded-full"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": "gray",
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          />
        </svg>
        <input
          className="w-full dark:text-gray-100 bg-gray-100 dark:bg-gray-700 focus:outline-none"
          inputMode="text"
          onChange={[Function]}
          onFocus={[Function]}
          onKeyPress={[Function]}
          onSubmit={[Function]}
          placeholder="Search..."
          type="search"
          value=""
        />
      </div>
    </div>
  </div>
</header>
`;
